src/source/Signaling/LwsApiCalls.c

```diff
@@ -1292,12 +1292,17 @@ PVOID reconnectHandler(PVOID args)
 
     CHK(pSignalingClient != NULL, STATUS_NULL_ARG);
 
+    // Indicate that we started
+    ATOMIC_STORE_BOOL(&pSignalingClient->reconnecterTracker.terminated, FALSE);
+
     // Await for the listener to clear
     MUTEX_LOCK(pSignalingClient->listenerTracker.lock);
     MUTEX_UNLOCK(pSignalingClient->listenerTracker.lock);
 
-    // Indicate that we started
-    ATOMIC_STORE_BOOL(&pSignalingClient->reconnecterTracker.terminated, FALSE);
+    // Exit immediately if we are shutting down in case we are getting terminated while we were waiting for the
+    // listener thread to terminate. The shutdown flag would have been checked prior kicking off the reconnect
+    // thread but there is a slight chance of a race condition.
+    CHK(!ATOMIC_LOAD_BOOL(&pSignalingClient->shutdown), retStatus);
```

Comment:
I am not confident this would fix the race condition mentioned in #861, at least for the usage in the sample code. Assuming create reconnectHandler thread takes super long time, the pSignalingClient is about to be freed due to refresh ICE error but not yet, so the pSignalingClient might be in a weird state, e.g. pSignalingClient is not NULL, but pSignalingClient->listenerTracker.lock is freed?

Reply:
You are correct. We have two options here

1) Try to set the terminated before creating the reconnect thread. This is actually pretty clean but need proper cleanup sequence in case the thread creation fails.
2) Use the thread trackers lock and the conditional variable to sequence the startup for the thread. This is actually more code than the first approach.

Will code up the first.